File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ls = new ArrayList<List<Integer>>();
        if(root==null) return ls;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        boolean leftToRight = true;
        while(!q.isEmpty()){
            int n=q.size();
            List<Integer> l = new ArrayList<>();
            for(int i=0;i<n;i++){
          
                if(q.peek().left !=null) q.offer(q.peek().left);
                if(q.peek().right !=null) q.offer(q.peek().right);
                l.add(q.poll().val);
            }
            if(leftToRight==false){
                Collections.reverse(l);
            }
            leftToRight = !leftToRight;
            ls.add(l);
        
        }
        return ls;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        
        vector<vector<int>>ans;
        if(root==NULL) return ans;
        queue<TreeNode*>q;
        q.push(root);
        bool lr= false;
        while(!q.empty()) {
            int n = q.size();
            vector<int> temp;
            for(int i=0;i<n;i++) {
                TreeNode *n = q.front();
                if(n->left) q.push(n->left);
                if(n->right) q.push(n->right);
                temp.push_back(n->val);
                q.pop();
            }
            if(lr) {
                reverse(temp.begin(),temp.end());
            }
            ans.push_back(temp);
            lr = !lr;
        }
        return ans;
    }
};